include/nanorange/ranges/istream_range.hpp

```diff
@@ -0,0 +1,123 @@
+// nanorange/ranges/istream_range.hpp
+//
+// Copyright (c) 2018 Tristan Brindle (tcbrindle at gmail dot com)
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef NANORANGE_RANGES_ISTREAM_RANGE_HPP_INCLUDED
+#define NANORANGE_RANGES_ISTREAM_RANGE_HPP_INCLUDED
+
+#include <nanorange/detail/iterator/traits.hpp>
+#include <nanorange/iterator/default_sentinel.hpp>
+
+#include <iosfwd>
+
+NANO_BEGIN_NAMESPACE
+
+template <typename T, typename CharT = char, typename Traits = std::char_traits<CharT>,
+          typename Distance = std::ptrdiff_t>
+class istream_range {
+public:
+
+    class iterator {
+    public:
+        using iterator_category = ranges::input_iterator_tag;
+        using difference_type = Distance;
+        using value_type = T;
+        using reference = const T&;
+        using pointer = const T*;
+        using char_type = CharT;
+        using traits = Traits;
+
+        constexpr iterator() = default;
+
+        explicit iterator(istream_range& rng)
+            : rng_(std::addressof(rng))
+        {}
+
+        reference operator*() const { return rng_->value_; }
+
+        pointer operator->() const { return std::addressof(rng_->value_); }
+
+        iterator& operator++()
+        {
+            rng_->next();
+            return *this;
+        }
+
+        iterator operator++(int)
+        {
+            auto tmp = *this;
+            this->operator++();
+            return tmp;
+        }
+
+        friend bool operator==(const iterator& lhs, const iterator& rhs)
+        {
+            return lhs.rng_ == rhs.rng_;
+        }
+
+        friend bool operator!=(const iterator& lhs, const iterator& rhs)
+        {
+            return !(lhs == rhs);
+        }
+
+        friend bool operator==(const iterator& lhs, default_sentinel)
+        {
+            return lhs.rng_->done();
+        }
+
+        friend bool operator!=(const iterator& lhs, default_sentinel rhs)
+        {
+            return !(lhs == rhs);
+        }
+
+        friend bool operator==(default_sentinel, const iterator& rhs)
+        {
+            return rhs.rng_->done();
+        }
+
+        friend bool operator!=(default_sentinel lhs, const iterator& rhs)
+        {
+            return !(lhs == rhs);
+        }
+
+    private:
+        istream_range* rng_ = nullptr;
+    };
+
+    using sentinel = default_sentinel;
+    using istream_type = std::basic_istream<CharT, Traits>;
+
+    constexpr istream_range() = default;
+
+    istream_range(istream_type& s)
+        : in_stream_(std::addressof(s))
+    {
+        next();
+    }
+
+    iterator begin() { return iterator{*this}; }
+
+    sentinel end() { return {}; }
+
+
+private:
+    void next()
+    {
+        *in_stream_ >> value_;
+        if (in_stream_->fail()) {
```

Comment:
Extraction operators are required (IIRC) to return a reference to the stream. `istream`s are explicitly convertible to `bool`, and the value returned from that conversion is `!fail()`. Putting all of that together, you can write this as:
```c++
if (!(*in_stream_ >> value_)) { 
    in_stream_ = nullptr; 
}
```
and save a function call.